Feature/best GitHub repos egypt

iahmedgamal opened the pull request:
# 📖 Description

Get the top GitHub repos from the BE and show them in a new page. 

## 🔄 Change Type


- [x] New feature (non-breaking change which adds functionality)

## 📷 Screenshots
[screen-capture (4).webm](https://github.com/egytech-fyi/egytech-fyi/assets/16765061/2862fdf4-8e92-4b8f-b064-7a8e70b48984)



## 🧪 How Has This Been Tested?
I tested it locally as  you can see in the video. 

## 🚧 Affected Parts

the header, 
I added a new page.

## 📋 Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

iahmedgamal committed:
top github repos

iahmedgamal committed:
eslint

iahmedgamal committed:
improvment

iahmedgamal committed:
eslint

iahmedgamal committed:
take only 10 topics for every repo

iahmedgamal committed:
display the language tag

iahmedgamal committed:
on click open the repo

BassemHalim commented:
Amazing work! looks great. 
In my opinion, I think the color of the stars is inconsistent with the style and might looks nicer in blue. maybe something like this:
![image](https://github.com/egytech-fyi/egytech-fyi/assets/20746657/dc647c4e-bd10-450a-9669-e78ff73551df)
Also green is used in other pages as an accent color so it might be nice to use somewhere in the design.

FathyMuhamed commented on src/components/NavBar.tsx:
```diff
@@ -26,6 +26,7 @@ export const Navbar = ({ handleOpenDrawer }: NavbarProps) => {
   }, [location.pathname])
 
   const isReport = location.pathname === '/report'
+  const twoToneColor = '#4c88ef'
```
i think can move to `src/constants/colors.ts`

iahmedgamal replied:
✔️ Done

FathyMuhamed commented on src/pages/Repos.tsx:
```diff
@@ -0,0 +1,42 @@
+import { useGetRepos } from '@services'
+
+export const Repos = () => {
+  const [isLoading, setIsLoading] = useState(true)
+
+  const { isError, data } = useGetRepos()
+
+  useEffect(() => {
+    if (data) {
+      setIsLoading(false)
+    }
+  }, [data])
```
```suggestion
  const { isError, data,isLoading } = useGetRepos()
```

iahmedgamal replied:
✔️ Done, I like it now it's cleaner

iahmedgamal committed:
fixed all comments

iahmedgamal commented:
> Amazing work! looks great. In my opinion, I think the color of the stars is inconsistent with the style and might looks nicer in blue. maybe something like this: ![image](https://private-user-images.githubusercontent.com/20746657/344424608-dc647c4e-bd10-450a-9669-e78ff73551df.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk4NDA5NDksIm5iZiI6MTcxOTg0MDY0OSwicGF0aCI6Ii8yMDc0NjY1Ny8zNDQ0MjQ2MDgtZGM2NDdjNGUtYmQxMC00NTBhLTk2NjktZTc4ZmY3MzU1MWRmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAxVDEzMzA0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThmMGNkNzNlMjIxNWFhMWFhMmY4MDhiZWYzNjEzZDExM2IxN2M1NmQ1MjAyNWRmNTNmZThmMThmNDRkMTczODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4hZTqhaKKkkLd7lqyRRdKaHYgKCkVvPmRncq0YdVNhA) Also green is used in other pages as an accent color so it might be nice to use somewhere in the design.

✔️ Done

iahmedgamal commented:
@MrViometal Can you have a look please for this PR?

MrViometal commented on src/components/NavBar.tsx:
```diff
@@ -42,15 +43,21 @@ export const Navbar = ({ handleOpenDrawer }: NavbarProps) => {
           className='navbar-menu-item'
           key='/dashboard'
-          icon={<DollarTwoTone twoToneColor='#4c88ef' />}>
+          icon={<DollarTwoTone twoToneColor={twoToneColor} />}>
```
`twoToneColor` is not descriptive, please rename it to a descriptive/distinguishable colour name

MrViometal commented on src/components/RepoCard.tsx:
```diff
@@ -0,0 +1,64 @@
+
+const { Title, Text } = Typography
+
+interface RepoCardProp {
```
```suggestion

interface RepoCardProps {
```

iahmedgamal replied:
Done ✅

MrViometal commented on src/components/RepoCard.tsx:
```diff
@@ -0,0 +1,64 @@
+  repo: GithubRepoContent
+}
+export const RepoCard = ({ repo }: RepoCardProp) => {
+  interface EmojiMap {
+    [key: string]: string
+  }
+  const emojiMap: EmojiMap = {
+    ':eyes:': '👀',
+    ':pencil2:': '✏️',
+  }
```
the interface and the object shouldn't live inside of RepoCard, it is bad practice in general to have interfaces and types defined inside of a component/function block

iahmedgamal replied:
Moved to `replaceShortcodeEmojis.ts` under `utils` folder

MrViometal replied:
I believe it should live under the `constants` folder, could you move it to keep the consistency 🙏🏼

MrViometal commented on src/components/RepoCard.tsx:
```diff
@@ -0,0 +1,64 @@
+    ':pencil2:': '✏️',
+  }
+
+  const convertShortcodesToUnicode = (text: string) => {
+    const regex = /(:\w+:)/g
+    return text.replace(regex, (match) => emojiMap[match] || match)
+  }
```
this function also doesn't necessarily need to be defined as a component method, can be defined outside of it

MrViometal commented on src/components/RepoCard.tsx:
```diff
@@ -0,0 +1,64 @@
+          <Title className='card-title' level={3}>
+            {repo.name}
+          </Title>
+          <Text className='card-description'>
+            {repo.description && convertShortcodesToUnicode(repo.description)}
+          </Text>
```
it doesn't make sense to have the node rendered if there is no description and is going to be childless

```suggestion
        {repo.description && ( 
            <Text className='card-description'>
              {convertShortcodesToUnicode(repo.description)}
            </Text>
         )}
```

iahmedgamal replied:
done.

MrViometal commented on src/components/RepoCard.tsx:
```diff
@@ -0,0 +1,64 @@
+          </Text>
+        </div>
+
+        {repo.topics.length === 0 && (
+          <Row justify='start' className='card-row'>
+            <RoundedTag tagName={repo.language} />
+          </Row>
+        )}
+
+        {repo.topics.length > 0 && (
+          <Row justify='start' className='card-row'>
+            {repo.topics.slice(0, 10).map((topic, index) => (
+              <RoundedTag key={index} tagName={topic} />
+            ))}
+          </Row>
+        )}
```
this logic can be beautifully encapsulated in a transformation function 
```suggestion
          <Row justify='start' className='card-row'>
          {transformRepoTopics(repo.topics).map((topic, index) => (
              <RoundedTag key={index} tagName={topic} />
            ))}
          </Row>
```

iahmedgamal replied:
Moved to `transformRepoTopics.ts` under `Transformers`

MrViometal commented on src/pages/Repos.tsx:
```diff
@@ -0,0 +1,33 @@
+      <Divider />
+      <div className='row-container'>
+        <Row gutter={{ xs: 8, sm: 16, md: 24, lg: 32 }} justify='center'>
+          {data &&
+            data.map((repo: GithubRepoContent) => (
```
nitpick

```suggestion
          {data?.map((repo: GithubRepoContent) => (
```

iahmedgamal replied:
done

MrViometal commented on src/styles/RepoCard.scss:
```diff
@@ -0,0 +1,56 @@
+.card-bg {
```
styles files have a naming convention `[fileName].styles.scss`, kindly make sure all style files submitted follow this convention

iahmedgamal replied:
Renamed from `RepoCard.scss` to  `RepoCard.styles.scss`

iahmedgamal committed:
clearer and more
concise logic within RepoCard component &&
Encapsulated topic transformation logic into utils.ts

iahmedgamal committed:
eslint

iahmedgamal committed:
rename a new style to have .styles.scss

iahmedgamal commented on src/components/NavBar.tsx:
```diff
@@ -42,15 +43,21 @@ export const Navbar = ({ handleOpenDrawer }: NavbarProps) => {
           className='navbar-menu-item'
           key='/dashboard'
-          icon={<DollarTwoTone twoToneColor='#4c88ef' />}>
+          icon={<DollarTwoTone twoToneColor={twoToneColor} />}>
```
Done, changed to `darkSkyBlue`

iahmedgamal committed:
updated the transformRepoTopics to include the
language

iahmedgamal committed:
eslint

iahmedgamal commented on src/utils/Transformers/transformRepoTopics.ts:
```diff
@@ -0,0 +1,9 @@
+    return [language]
+  } else {
+    return topics.slice(0, MAX_TOPICS)
+  }
```
Resolved 🫡

MrViometal commented on src/components/RepoCard.tsx:
```diff
@@ -0,0 +1,46 @@
+import '@styles/RepoCard.styles.scss'
+import { RoundedTag } from './RoundedTag'
```
should be exported from `@components`

MrViometal commented on src/components/RepoCard.tsx:
```diff
@@ -0,0 +1,46 @@
+    <a href={repo.html_url} target='_blank' rel='noopener noreferrer' className='repo-card'>
+      <div className='card-bg'>
+        <Row justify='space-between' align='middle' className='card-row'>
+          <Col span={3}>{repo.language && <RoundedTag tagName={repo.language} />}</Col>
```
if no repo.language, "false" will be rendered instead, which is unwanted, a ternary operator: 
```suggestion
          {repo.language ? 
                <Col span={3}>
                    <RoundedTag tagName={repo.language} />
                </Col>
                : null
          }
```

or a short-circuit 
```suggestion
          {repo.language && 
                <Col span={3}>
                    <RoundedTag tagName={repo.language} />
                </Col>
           }
```
to determine whether to show the whole Col or not in the first place would solve this problem

MrViometal replied:
since we already short-circuited other datapoints in this component, let's keep the consistency and use it instead of a ternary operator

iahmedgamal replied:
I did it for  `language` and  `stargazers_count` as well

MrViometal commented on src/pages/Repos.tsx:
```diff
@@ -0,0 +1,32 @@
+  if (isLoading) return <Spin size='large' style={{ margin: 'auto' }} />
+
+  return (
+    <div style={{ width: '100%' }}>
```
it is bad design to have inline styles, kindly use className instead

iahmedgamal replied:
Refactored this to have `full-width` class

iahmedgamal committed:
Removed inline style from the repo &&
Short-circuit for language and stargazers_count &&
Imported RoundedTag from @components
